### DISCUSSION TITLE
Spacing after a glyph followed by a mark

### DISCUSSION THREAD

**Discussion by goldenadav** [fontforge/fontforge#5474] [General]
Hello! I am designing a special Hebrew font for my company's use. I changed an existing mark (unicode U+05BF) and added anchors and positioned them on the Hebrew letters with GPOS Lookup Mark to Base. Now, I noticed that since the mark is positioned in top-front of the letter, it will be too close to the next letter. So my quest is to add spacing after letters that are followed by this mark (maybe even a different spacing for each other letter). I am not sure how to do it, even after a lot of Googling and ChatGPTing). Hope that someone here can help me :)
![image](https://github.com/user-attachments/assets/69fcce35-13f1-4e3b-b76d-55cacd7eadf8)
You can see in the image (Right-to-left) that the mark (marked by 2, it's not a normal apostrophe) comes after the letter marked by 1. The letter marked by 3 comes right after and is too close to the previous letter with the mark. I hope that you understand what I mean. How can I solve it?

**Comment by iorsh** [comment]
Normally RAFE should not be replaced with GERESH, but if you really need, you can try one of the following:
1. Use GSUB lookup to replace u+05BF with u+05F3, and use kerning to improve positioning
2. Designate u+05BF explicitly as base glyph and give it non-zero width. This could be problematic, as some older renderers ignore this designation.

**Reply by goldenadav** [reply]
What would be the right way to replace Glyph + Mark to a new Glyph? (e.g. replacing Gimel + RAFE to my custom uni0001)

**Reply by iorsh** [reply]
Ligature substitution, I guess. 
https://fontforge.org/docs/ui/dialogs/lookups.html#lookups-basic-subs

**Comment by goldenadav** [comment]
I will try to explain the issue in another way. My main mission is to modify an existing Hebrew font in order to create a version suitable for transliterated Arabic using Hebrew letters. For instance, I would like to add a Hamza mark above the letter Alef and be able to place a Shadda mark on top of the Hamza. 

My first approach was to try adding Ligatures from a few Hebrew letters + Hebrew marks and change them to custom glyphs (e.g /uni05D0 (Alef) + /uni05AB => /uni0001, a glyph of Alef with Hamza). The Ligature worked, but the problem here was that, in MS Word, when I tried to add another mark to the ligature, it broke the ligature back to it's components.

My second approach was to to modify the existing Hebrew marks (e.g I changed /uni05AB to the shape of Hamza) and then create new anchors for them. Each of the marks worked fine, but this time I could not get the marks to position one on top of the other with Mark to Mark. I noticed that they do position when I write them without a base glyph, but when I added a base glyph both of them were positioned on top of the base and not one according to the other (If you want I can add screens of that). In one of my tries in this approach, I noticed that when I add the first mark 2 times on top of the base, and then add the second mark, they do position correctly, but of course its not really helping me...

Anyways, I am left clueless and hope that you can help me. Do you have any suggestions on how to achieve my goal? I would greatly appreciate any insights you could share, and I thank you in advance for your help.

**Reply by iorsh** [reply]
If you have a limited set of marked and double-marked letters, it would be easiest to create composite glyphs for all of them and to define double and triple ligatures as appropriate. Please note the order of rule application - put triple ligatures first and double after them.

Messing with `mkmk` feature is hard. I spent much more that a few hours on it, and my solution still breaks in Windows and works only with Harfbuzz.

You can check the rules in e.g. "David CLM" and see how they can be adapted to your case: https://sourceforge.net/projects/culmus/files/culmus/0.140/culmus-src-0.140.tar.gz/download

I can't think right now of a simple font which does `mkmk` feature right. Maybe [Arabeyes](https://www.arabeyes.org/Khotot) does something like that, but I didn't check that in depth.

**Reply by goldenadav** [reply]
I considered creating a 3-way ligature last week, but I thought it might be a bit too complex.

I just tested it and noticed that while the ligature works for `two base characters and one mark`, it doesn’t seem to work with `a single base and two marks`.

I’ll keep thinking about how to handle this, but if you have any further insights, I’d love to hear them!

Thanks!

**Comment by iorsh** [comment]
Mark order is sometimes normalized, i.e. swapped according to some predefined Unicode rules. Try to encode both Aleph+Shadda+Hamza and Aleph+Hamza+Shadda.

**Reply by goldenadav** [reply]
You are right! It really solved the 3-way ligature with 2 marks, but now I have a different problem because adding the regular Hebrew marks breaks the ligature

**Reply by iorsh** [reply]
Well, this is getting more and more complicated, I must admit. My first impression was that you are trying to use Hebrew consonants with a limited set of Arabic vowels, but now it turns out there are Hebrew marks too. Can you please, just for reference, screenshot a few most complex cases of what you are trying to achieve? It would be easier to consider methods when the entire picture is known.

**Reply by goldenadav** [reply]
Here you can see a few examples of Hebrew letters with complicated marks on top of them
![image](https://github.com/user-attachments/assets/52b4303b-8fdd-490b-ac58-ff604dd5066a)

**Comment by goldenadav** [comment]
I have another idea I’d like to try: creating ligatures for the marks themselves, e.g `Hamza+Shadda=HamzaShadda`. I’ll test it now and follow up with the results.

Edit: I have a feeling that it will not work because of the same Unicode mark swapping that you mentioned above. But worth a try

**Reply by goldenadav** [reply]
As I thought - adding another mark breaks the ligature

**Reply by iorsh** [reply]
You might need to take a look at [Canonical_ordering](https://en.wikipedia.org/wiki/Unicode_equivalence#Canonical_ordering). In general, Unicode doesn't address mixes of marks from different languages, so canonical reordering could be rather confusing in that case. Still, if you have a limited number of mark combinations, ligatures could be the way to go. You would still need to figure what is the canonical order for all your marks.

Please note that upper and lower mark don't need to be combined together. You can combine upper marks, combine lower marks, and attach each ligature to the base. As far as I remember, you can direct mark position lookup to ignore other marks.
